model_api/python/model_api/models/visual_prompting.py

```diff
@@ -0,0 +1,757 @@
+        """
+        Executes `learn` stage of SAM ZSL pipeline.
+
+        Reference features are updated according to newly arrived prompts. Features corresponding to the same labels are overridden during
+        consequent learn() calls.
+
+        Args:
+            image (np.ndarray): HWC-shaped image
+            boxes (Prompt | None, optional): Prompt containing bounding boxes (in XYXY torchvision format) and their labels (ints, one per box). Defaults to None.
+            points (Prompt | None, optional): Prompt containing bounding boxes (in XY format) and their labels (ints, one per point). Defaults to None.
+            reset_features (bool, optional): Forces learning from scratch. Defaults to False.
+
+        Returns:
+            tuple[VisualPromptingFeatures, np.ndarray]: return values are the updated VPT reference features and reference masks.
+            The shape of the reference mask is N_labels x H x W, where H and W are the same as in the input image.
+        """
+
+        if boxes is None and points is None:
+            raise RuntimeError("boxes or points prompts are required for learning")
+
+        if reset_features or not self.has_reference_features():
+            self.reset_reference_info()
+
+        processed_prompts = self.decoder.preprocess(
+            {
+                "bboxes": boxes.data if boxes else None,
+                "points": points.data if points else None,
+                "labels": {"bboxes": boxes.labels if boxes else None,
+                           "points": points.labels if points else None},
+                "orig_size": image.shape[:2],
+            },
+        )
+
+        processed_prompts_w_labels = self._gather_prompts_with_labels(processed_prompts)
+        largest_label: int = max([int(p) for p in processed_prompts_w_labels] + [0])
+        self._expand_reference_info(largest_label)
+
+        original_shape = np.array(image.shape[:2])
+
+        # forward image encoder
+        image_embeddings = self.encoder(image)
+        processed_embedding = image_embeddings.squeeze().transpose(1, 2, 0)
+
+        # get reference masks
+        ref_masks: np.ndarray = np.zeros(
+            (largest_label + 1, *original_shape), dtype=np.uint8
+        )
+        for label, input_prompts in processed_prompts_w_labels.items():
+            ref_mask: np.ndarray = np.zeros(original_shape, dtype=np.uint8)
+            for inputs_decoder in input_prompts:
+                inputs_decoder.pop("label")
+                if "point_coords" in inputs_decoder:
+                    # bboxes and points
+                    inputs_decoder["image_embeddings"] = image_embeddings
+                    prediction = self._predict_masks(
+                        inputs_decoder, original_shape, is_cascade=self._is_cascade
+                    )
+                    masks = prediction["upscaled_masks"]
+                else:
+                    raise RuntimeError("Prompts other than points are not supported")
```

Comment:
@maxxgx Do you consider a case with polygons (mask or scribble generated from Geti)?
If so, we can refer this implementation I did for demo last year.
https://github.com/sungchul2/training_extensions/blob/a5f869a558f90afda500495cde5431679be2f023/src/otx/algorithms/visual_prompting/adapters/torch/processor/zero_shot_processor.py#L576-L579

Reply:
Indeed, we also need polygon type prompts. Thanks for the reminder!

Reply:
@sungchul2 could you prepare a slide with requirements for polygon annotation and send it to me, so we can discuss it?